celery/worker/consumer/delayed_delivery.py

```diff
@@ -9,29 +17,103 @@
 __all__ = ('DelayedDelivery',)
 
 logger = get_logger(__name__)
+debug, info = logger.debug, logger.info
 
 
 class DelayedDelivery(bootsteps.StartStopStep):
-    """This bootstep declares native delayed delivery queues and exchanges and binds all queues to them"""
+    """Bootstep that configures native delayed delivery for quorum queues.
+
+    This component declares the necessary exchanges and queues for native
+    delayed delivery and binds all worker queues to them. When enabled,
+    tasks with ETA/countdown will be delivered by the broker itself
+    rather than being held in worker memory.
+
+    Note:
+        This bootstep is only active when quorum queues are detected.
+        Direct exchanges are not supported by native delayed delivery.
+    """
+
     requires = (Tasks,)
 
     def include_if(self, c):
-        return detect_quorum_queues(c.app, c.app.connection_for_write().transport.driver_type)[0]
+        """Decide if this bootstep should be included.
+
+        Only include this bootstep if quorum queues are detected in the broker.
+
+        Arguments:
+            c (Consumer): Consumer instance.
+
+        Returns:
+            bool: True if quorum queues are detected, False otherwise.
+        """
+        return detect_quorum_queues(
+            c.app, c.app.connection_for_write().transport.driver_type)[0]
 
     def start(self, c: Consumer):
+        """Set up native delayed delivery for quorum queues.
+
+        This configures the necessary exchanges and bindings to enable
+        broker-based delivery of delayed tasks.
+
+        Arguments:
+            c (Consumer): Consumer instance.
+        """
+        info('setting up native delayed delivery for quorum queues')
+        self._setup_delayed_delivery(c)
+        info('native delayed delivery setup completed')
+
+    def _setup_delayed_delivery(self, c: Consumer):
+        """Set up delayed delivery exchanges and queues for each broker.
+
+        For each broker in a multi-broker configuration, this method:
+        1. Creates a connection
+        2. Declares the delayed delivery exchanges and queues
+        3. Binds the worker queues to these exchanges
+
+        Error handling ensures failures with one broker don't affect others.
+
+        Arguments:
+            c (Consumer): Consumer instance.
+        """
         app: Celery = c.app
 
         for broker_url in app.conf.broker_url.split(';'):
```

Comment:
There was also a bug in this code that was accepted as a known issue but was never documented. I left it for the agents to find it out naturally.
It was about error handling when there's a valid error during this loop @adrinjalali.